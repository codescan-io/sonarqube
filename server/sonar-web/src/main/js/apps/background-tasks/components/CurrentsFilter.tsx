/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Checkbox from '../../../components/controls/Checkbox';
import { translate } from '../../../helpers/l10n';
import { CURRENTS } from '../constants';

interface CurrentsFilterProps {
  value?: string;
  id: string;
  onChange: (value: string) => void;
}

export default function CurrentsFilter(props: CurrentsFilterProps) {
  const { id, value, onChange } = props;
  const checked = value === CURRENTS.ONLY_CURRENTS;

  const handleChange = React.useCallback(
    (value: boolean) => {
      const newValue = value ? CURRENTS.ONLY_CURRENTS : CURRENTS.ALL;
      onChange(newValue);
    },
    [onChange]
  );

  return (
    <div className="bt-search-form-field">
      <Checkbox id={id} checked={checked} onCheck={handleChange}>
        <span className="little-spacer-left">{translate('yes')}</span>
      </Checkbox>
    </div>
  );
}
