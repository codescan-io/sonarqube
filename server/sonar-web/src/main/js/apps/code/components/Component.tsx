/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import classNames from 'classnames';
import * as React from 'react';
import { withScrollTo } from '../../../components/hoc/withScrollTo';
import DateFromNow from '../../../components/intl/DateFromNow';
import { WorkspaceContext } from '../../../components/workspace/context';
import { BranchLike } from '../../../types/branch-like';
import { ComponentQualifier } from '../../../types/component';
import { ComponentMeasure as TypeComponentMeasure, Metric } from '../../../types/types';
import ComponentMeasure from './ComponentMeasure';
import ComponentName from './ComponentName';
import ComponentPin from './ComponentPin';

interface Props {
  branchLike?: BranchLike;
  canBePinned?: boolean;
  canBrowse?: boolean;
  component: TypeComponentMeasure;
  hasBaseComponent: boolean;
  isBaseComponent?: boolean;
  metrics: Metric[];
  previous?: TypeComponentMeasure;
  rootComponent: TypeComponentMeasure;
  selected?: boolean;
  newCodeSelected?: boolean;
  showAnalysisDate?: boolean;
}

export class Component extends React.PureComponent<Props> {
  render() {
    const {
      branchLike,
      canBePinned = true,
      canBrowse = false,
      component,
      hasBaseComponent,
      isBaseComponent = false,
      metrics,
      previous,
      rootComponent,
      selected = false,
      newCodeSelected,
      showAnalysisDate,
    } = this.props;

    const isFile =
      component.qualifier === ComponentQualifier.File ||
      component.qualifier === ComponentQualifier.TestFile;

    return (
      <tr className={classNames({ selected, 'current-folder': isBaseComponent })}>
        {canBePinned && (
          <td className="thin nowrap">
            {isFile && (
              <WorkspaceContext.Consumer>
                {({ openComponent }) => (
                  <ComponentPin
                    branchLike={branchLike}
                    component={component}
                    openComponent={openComponent}
                  />
                )}
              </WorkspaceContext.Consumer>
            )}
          </td>
        )}
        <td className="code-name-cell">
          <div className="display-flex-center">
            {hasBaseComponent && <div className="code-child-component-icon" />}
            <ComponentName
              branchLike={branchLike}
              canBrowse={canBrowse}
              component={component}
              previous={previous}
              rootComponent={rootComponent}
              unclickable={isBaseComponent}
              newCodeSelected={newCodeSelected}
            />
          </div>
        </td>

        {metrics.map((metric) => (
          <td className="text-center" key={metric.key}>
            <ComponentMeasure component={component} metric={metric} />
          </td>
        ))}

        {showAnalysisDate && isBaseComponent && <td />}

        {showAnalysisDate && !isBaseComponent && (
          <td className="text-center">
            {component.analysisDate ? <DateFromNow date={component.analysisDate} /> : 'â€”'}
          </td>
        )}
      </tr>
    );
  }
}

export default withScrollTo(Component);
