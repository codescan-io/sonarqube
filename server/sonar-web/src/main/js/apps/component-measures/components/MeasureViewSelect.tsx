/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { components, OptionProps, SingleValueProps } from 'react-select';
import Select from '../../../components/controls/Select';
import ListIcon from '../../../components/icons/ListIcon';
import TreeIcon from '../../../components/icons/TreeIcon';
import TreemapIcon from '../../../components/icons/TreemapIcon';
import { translate } from '../../../helpers/l10n';
import { MeasurePageView } from '../../../types/measures';
import { Metric } from '../../../types/types';
import { hasList, hasTree, hasTreemap } from '../utils';

interface Props {
  className?: string;
  metric: Metric;
  handleViewChange: (view: MeasurePageView) => void;
  view: MeasurePageView;
}

interface ViewOption {
  icon: JSX.Element;
  label: string;
  value: string;
}

export default class MeasureViewSelect extends React.PureComponent<Props> {
  getOptions = () => {
    const { metric } = this.props;
    const options: ViewOption[] = [];
    if (hasTree(metric.key)) {
      options.push({
        icon: <TreeIcon />,
        label: translate('component_measures.tab.tree'),
        value: 'tree',
      });
    }
    if (hasList(metric.key)) {
      options.push({
        icon: <ListIcon />,
        label: translate('component_measures.tab.list'),
        value: 'list',
      });
    }
    if (hasTreemap(metric.key, metric.type)) {
      options.push({
        icon: <TreemapIcon />,
        label: translate('component_measures.tab.treemap'),
        value: 'treemap',
      });
    }
    return options;
  };

  handleChange = (option: ViewOption) => {
    return this.props.handleViewChange(option.value as MeasurePageView);
  };

  renderOption = (props: OptionProps<ViewOption, false>) => (
    <components.Option {...props} className="display-flex-center">
      {props.data.icon}
      <span className="little-spacer-left">{props.data.label}</span>
    </components.Option>
  );

  renderValue = (props: SingleValueProps<ViewOption>) => (
    <components.SingleValue {...props} className="display-flex-center">
      {props.data.icon}
      <span className="little-spacer-left">{props.data.label}</span>
    </components.SingleValue>
  );

  render() {
    const { className, view } = this.props;
    const options = this.getOptions();

    return (
      <Select
        aria-labelledby="measures-view-selection-label"
        blurInputOnSelect={true}
        className={className}
        onChange={this.handleChange}
        components={{
          Option: this.renderOption,
          SingleValue: this.renderValue,
        }}
        options={options}
        isSearchable={false}
        value={options.find((o) => o.value === view)}
      />
    );
  }
}
