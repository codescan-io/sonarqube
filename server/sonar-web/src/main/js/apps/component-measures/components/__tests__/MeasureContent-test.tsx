/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { getComponentTree } from '../../../../api/components';
import { mockComponentMeasure } from '../../../../helpers/mocks/component';
import { mockRouter } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import MeasureContent from '../MeasureContent';

jest.mock('../../../../helpers/scrolling', () => ({
  scrollToElement: jest.fn(),
}));

jest.mock('../../../../api/components', () => {
  const { mockComponentMeasure } = jest.requireActual('../../../../helpers/mocks/component');
  return {
    getComponentTree: jest.fn().mockResolvedValue({
      paging: { pageIndex: 1, pageSize: 500, total: 2 },
      baseComponent: mockComponentMeasure(),
      components: [mockComponentMeasure(true)],
      metrics: [
        {
          bestValue: '0',
          custom: false,
          description: 'Bugs',
          domain: 'Reliability',
          hidden: false,
          higherValuesAreBetter: false,
          key: 'bugs',
          name: 'Bugs',
          qualitative: true,
          type: 'INT',
        },
      ],
    }),
  };
});

jest.mock('../../../../api/measures', () => ({
  getMeasures: jest.fn().mockResolvedValue([{ metric: 'bugs', value: '12', bestValue: false }]),
}));

const METRICS = {
  bugs: { id: '1', key: 'bugs', type: 'INT', name: 'Bugs', domain: 'Reliability' },
};

const WINDOW_HEIGHT = 800;
const originalHeight = window.innerHeight;

beforeEach(() => {
  jest.clearAllMocks();
});

beforeAll(() => {
  Object.defineProperty(window, 'innerHeight', {
    writable: true,
    configurable: true,
    value: WINDOW_HEIGHT,
  });
});

afterAll(() => {
  Object.defineProperty(window, 'innerHeight', {
    writable: true,
    configurable: true,
    value: originalHeight,
  });
});

it('should render correctly for a project', async () => {
  const wrapper = shallowRender();
  expect(wrapper.type()).toBeNull();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should render correctly when asc prop is defined', async () => {
  const wrapper = shallowRender({ asc: true });
  expect(wrapper.type()).toBeNull();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should render correctly when view prop is tree', async () => {
  const wrapper = shallowRender({ view: 'tree' });
  expect(wrapper.type()).toBeNull();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should render correctly for a file', async () => {
  (getComponentTree as jest.Mock).mockResolvedValueOnce({
    paging: { pageIndex: 1, pageSize: 500, total: 0 },
    baseComponent: mockComponentMeasure(true),
    components: [],
    metrics: [METRICS.bugs],
  });
  const wrapper = shallowRender();
  expect(wrapper.type()).toBeNull();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should test fetchMoreComponents to work correctly', async () => {
  (getComponentTree as jest.Mock).mockResolvedValueOnce({
    paging: { pageIndex: 12, pageSize: 500, total: 0 },
    baseComponent: mockComponentMeasure(false),
    components: [],
    metrics: [METRICS.bugs],
  });
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  wrapper.instance().fetchMoreComponents();
  expect(wrapper).toMatchSnapshot();
});

it('should test getComponentRequestParams response for different arguments', () => {
  const wrapper = shallowRender({ asc: false });
  const metric = {
    direction: -1,
    key: 'new_reliability_rating',
  };
  const reqParamsList = {
    metricKeys: ['new_reliability_rating'],
    opts: {
      additionalFields: 'metrics',
      asc: true,
      metricPeriodSort: 1,
      metricSort: 'new_reliability_rating',
      metricSortFilter: 'withMeasuresOnly',
      ps: 500,
      s: 'metricPeriod',
    },
    strategy: 'leaves',
  };
  expect(wrapper.instance().getComponentRequestParams('list', metric, { asc: true })).toEqual(
    reqParamsList
  );
  // when options.asc is not passed the opts.asc will take the default value
  reqParamsList.opts.asc = false;
  expect(wrapper.instance().getComponentRequestParams('list', metric, {})).toEqual(reqParamsList);

  const reqParamsTreeMap = {
    metricKeys: ['new_reliability_rating', 'new_lines'],
    opts: {
      additionalFields: 'metrics',
      asc: true,
      metricPeriodSort: 1,
      metricSort: 'new_lines',
      metricSortFilter: 'withMeasuresOnly',
      ps: 500,
      s: 'metricPeriod',
    },
    strategy: 'children',
  };
  expect(wrapper.instance().getComponentRequestParams('treemap', metric, { asc: true })).toEqual(
    reqParamsTreeMap
  );
  // when options.asc is not passed the opts.asc will take the default value
  reqParamsTreeMap.opts.asc = false;
  expect(wrapper.instance().getComponentRequestParams('treemap', metric, {})).toEqual(
    reqParamsTreeMap
  );

  const reqParamsTree = {
    metricKeys: ['new_reliability_rating'],
    opts: {
      additionalFields: 'metrics',
      asc: false,
      ps: 500,
      s: 'qualifier,name',
    },
    strategy: 'children',
  };
  expect(wrapper.instance().getComponentRequestParams('tree', metric, { asc: false })).toEqual(
    reqParamsTree
  );
  // when options.asc is not passed the opts.asc will take the default value
  reqParamsTree.opts.asc = true;
  expect(wrapper.instance().getComponentRequestParams('tree', metric, {})).toEqual(reqParamsTree);
});

function shallowRender(props: Partial<MeasureContent['props']> = {}) {
  return shallow<MeasureContent>(
    <MeasureContent
      metrics={METRICS}
      requestedMetric={{ direction: 1, key: 'bugs' }}
      rootComponent={mockComponentMeasure()}
      router={mockRouter()}
      updateQuery={jest.fn()}
      view="list"
      {...props}
    />
  );
}
