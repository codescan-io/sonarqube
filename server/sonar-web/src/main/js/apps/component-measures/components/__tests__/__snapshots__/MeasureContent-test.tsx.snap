// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly for a file 1`] = `
<div
  className="layout-page-main no-outline"
>
  <A11ySkipTarget
    anchor="measures_main"
  />
  <div
    className="layout-page-header-panel layout-page-main-header"
  >
    <div
      className="layout-page-header-panel-inner layout-page-main-header-inner"
    >
      <div
        className="layout-page-main-inner"
      >
        <MeasureContentHeader
          left={
            <Breadcrumbs
              backToFirst={true}
              className="text-ellipsis flex-1"
              component={
                {
                  "key": "foo:src/index.tsx",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "1",
                    },
                  ],
                  "name": "index.tsx",
                  "path": "src/index.tsx",
                  "qualifier": "FIL",
                }
              }
              handleSelect={[Function]}
              rootComponent={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
            />
          }
          right={
            <div
              className="display-flex-center"
            />
          }
        />
      </div>
    </div>
  </div>
  <div
    className="layout-page-main-inner measure-details-content"
  >
    <MeasureHeader
      component={
        {
          "key": "foo:src/index.tsx",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "1",
            },
          ],
          "name": "index.tsx",
          "path": "src/index.tsx",
          "qualifier": "FIL",
        }
      }
      measureValue="12"
      metric={
        {
          "direction": 1,
          "domain": "Reliability",
          "id": "1",
          "key": "bugs",
          "name": "Bugs",
          "type": "INT",
        }
      }
    />
    <div
      className="measure-details-viewer"
    >
      <SourceViewer
        component="foo:src/index.tsx"
        displayAllIssues={false}
        displayIssueLocationsCount={true}
        displayIssueLocationsLink={true}
        displayLocationMarkers={true}
        metricKey="bugs"
      />
    </div>
  </div>
</div>
`;

exports[`should render correctly for a project 1`] = `
<div
  className="layout-page-main no-outline"
>
  <A11ySkipTarget
    anchor="measures_main"
  />
  <div
    className="layout-page-header-panel layout-page-main-header"
  >
    <div
      className="layout-page-header-panel-inner layout-page-main-header-inner"
    >
      <div
        className="layout-page-main-inner"
      >
        <MeasureContentHeader
          left={
            <Breadcrumbs
              backToFirst={true}
              className="text-ellipsis flex-1"
              component={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
              handleSelect={[Function]}
              rootComponent={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
            />
          }
          right={
            <div
              className="display-flex-center"
            >
              <React.Fragment>
                <div
                  id="measures-view-selection-label"
                >
                  component_measures.view_as
                </div>
                <MeasureViewSelect
                  className="measure-view-select spacer-left big-spacer-right"
                  handleViewChange={[Function]}
                  metric={
                    {
                      "bestValue": "0",
                      "custom": false,
                      "description": "Bugs",
                      "direction": 1,
                      "domain": "Reliability",
                      "hidden": false,
                      "higherValuesAreBetter": false,
                      "key": "bugs",
                      "name": "Bugs",
                      "qualitative": true,
                      "type": "INT",
                    }
                  }
                  view="list"
                />
                <PageActions
                  componentQualifier="TRK"
                  showShortcuts={true}
                  total={2}
                />
              </React.Fragment>
            </div>
          }
        />
      </div>
    </div>
  </div>
  <div
    className="layout-page-main-inner measure-details-content"
  >
    <MeasureHeader
      component={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      measureValue="12"
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
    />
    <FilesView
      components={
        [
          {
            "key": "foo:src/index.tsx",
            "leak": undefined,
            "measures": [
              {
                "bestValue": false,
                "leak": undefined,
                "metric": {
                  "domain": "Reliability",
                  "id": "1",
                  "key": "bugs",
                  "name": "Bugs",
                  "type": "INT",
                },
                "value": "1",
              },
            ],
            "name": "index.tsx",
            "path": "src/index.tsx",
            "qualifier": "FIL",
            "value": "1",
          },
        ]
      }
      defaultShowBestMeasures={false}
      fetchMore={[Function]}
      handleOpen={[Function]}
      handleSelect={[Function]}
      loadingMore={false}
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
      metrics={
        {
          "bugs": {
            "domain": "Reliability",
            "id": "1",
            "key": "bugs",
            "name": "Bugs",
            "type": "INT",
          },
        }
      }
      paging={
        {
          "pageIndex": 1,
          "pageSize": 500,
          "total": 2,
        }
      }
      rootComponent={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      view="list"
    />
  </div>
</div>
`;

exports[`should render correctly when asc prop is defined 1`] = `
<div
  className="layout-page-main no-outline"
>
  <A11ySkipTarget
    anchor="measures_main"
  />
  <div
    className="layout-page-header-panel layout-page-main-header"
  >
    <div
      className="layout-page-header-panel-inner layout-page-main-header-inner"
    >
      <div
        className="layout-page-main-inner"
      >
        <MeasureContentHeader
          left={
            <Breadcrumbs
              backToFirst={true}
              className="text-ellipsis flex-1"
              component={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
              handleSelect={[Function]}
              rootComponent={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
            />
          }
          right={
            <div
              className="display-flex-center"
            >
              <React.Fragment>
                <div
                  id="measures-view-selection-label"
                >
                  component_measures.view_as
                </div>
                <MeasureViewSelect
                  className="measure-view-select spacer-left big-spacer-right"
                  handleViewChange={[Function]}
                  metric={
                    {
                      "bestValue": "0",
                      "custom": false,
                      "description": "Bugs",
                      "direction": 1,
                      "domain": "Reliability",
                      "hidden": false,
                      "higherValuesAreBetter": false,
                      "key": "bugs",
                      "name": "Bugs",
                      "qualitative": true,
                      "type": "INT",
                    }
                  }
                  view="list"
                />
                <PageActions
                  componentQualifier="TRK"
                  showShortcuts={true}
                  total={2}
                />
              </React.Fragment>
            </div>
          }
        />
      </div>
    </div>
  </div>
  <div
    className="layout-page-main-inner measure-details-content"
  >
    <MeasureHeader
      component={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      measureValue="12"
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
    />
    <FilesView
      components={
        [
          {
            "key": "foo:src/index.tsx",
            "leak": undefined,
            "measures": [
              {
                "bestValue": false,
                "leak": undefined,
                "metric": {
                  "domain": "Reliability",
                  "id": "1",
                  "key": "bugs",
                  "name": "Bugs",
                  "type": "INT",
                },
                "value": "1",
              },
            ],
            "name": "index.tsx",
            "path": "src/index.tsx",
            "qualifier": "FIL",
            "value": "1",
          },
        ]
      }
      defaultShowBestMeasures={true}
      fetchMore={[Function]}
      handleOpen={[Function]}
      handleSelect={[Function]}
      loadingMore={false}
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
      metrics={
        {
          "bugs": {
            "domain": "Reliability",
            "id": "1",
            "key": "bugs",
            "name": "Bugs",
            "type": "INT",
          },
        }
      }
      paging={
        {
          "pageIndex": 1,
          "pageSize": 500,
          "total": 2,
        }
      }
      rootComponent={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      view="list"
    />
  </div>
</div>
`;

exports[`should render correctly when view prop is tree 1`] = `
<div
  className="layout-page-main no-outline"
>
  <A11ySkipTarget
    anchor="measures_main"
  />
  <div
    className="layout-page-header-panel layout-page-main-header"
  >
    <div
      className="layout-page-header-panel-inner layout-page-main-header-inner"
    >
      <div
        className="layout-page-main-inner"
      >
        <MeasureContentHeader
          left={
            <Breadcrumbs
              backToFirst={false}
              className="text-ellipsis flex-1"
              component={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
              handleSelect={[Function]}
              rootComponent={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
            />
          }
          right={
            <div
              className="display-flex-center"
            >
              <React.Fragment>
                <div
                  id="measures-view-selection-label"
                >
                  component_measures.view_as
                </div>
                <MeasureViewSelect
                  className="measure-view-select spacer-left big-spacer-right"
                  handleViewChange={[Function]}
                  metric={
                    {
                      "bestValue": "0",
                      "custom": false,
                      "description": "Bugs",
                      "direction": 1,
                      "domain": "Reliability",
                      "hidden": false,
                      "higherValuesAreBetter": false,
                      "key": "bugs",
                      "name": "Bugs",
                      "qualitative": true,
                      "type": "INT",
                    }
                  }
                  view="tree"
                />
                <PageActions
                  componentQualifier="TRK"
                  showShortcuts={true}
                  total={2}
                />
              </React.Fragment>
            </div>
          }
        />
      </div>
    </div>
  </div>
  <div
    className="layout-page-main-inner measure-details-content"
  >
    <MeasureHeader
      component={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      measureValue="12"
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
    />
    <FilesView
      components={
        [
          {
            "key": "foo:src/index.tsx",
            "leak": undefined,
            "measures": [
              {
                "bestValue": false,
                "leak": undefined,
                "metric": {
                  "domain": "Reliability",
                  "id": "1",
                  "key": "bugs",
                  "name": "Bugs",
                  "type": "INT",
                },
                "value": "1",
              },
            ],
            "name": "index.tsx",
            "path": "src/index.tsx",
            "qualifier": "FIL",
            "value": "1",
          },
        ]
      }
      defaultShowBestMeasures={true}
      fetchMore={[Function]}
      handleOpen={[Function]}
      handleSelect={[Function]}
      loadingMore={false}
      metric={
        {
          "bestValue": "0",
          "custom": false,
          "description": "Bugs",
          "direction": 1,
          "domain": "Reliability",
          "hidden": false,
          "higherValuesAreBetter": false,
          "key": "bugs",
          "name": "Bugs",
          "qualitative": true,
          "type": "INT",
        }
      }
      metrics={
        {
          "bugs": {
            "domain": "Reliability",
            "id": "1",
            "key": "bugs",
            "name": "Bugs",
            "type": "INT",
          },
        }
      }
      paging={
        {
          "pageIndex": 1,
          "pageSize": 500,
          "total": 2,
        }
      }
      rootComponent={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      view="tree"
    />
  </div>
</div>
`;

exports[`should test fetchMoreComponents to work correctly 1`] = `
<div
  className="layout-page-main no-outline"
>
  <A11ySkipTarget
    anchor="measures_main"
  />
  <div
    className="layout-page-header-panel layout-page-main-header"
  >
    <div
      className="layout-page-header-panel-inner layout-page-main-header-inner"
    >
      <div
        className="layout-page-main-inner"
      >
        <MeasureContentHeader
          left={
            <Breadcrumbs
              backToFirst={true}
              className="text-ellipsis flex-1"
              component={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
              handleSelect={[Function]}
              rootComponent={
                {
                  "key": "foo",
                  "measures": [
                    {
                      "bestValue": false,
                      "metric": "bugs",
                      "value": "12",
                    },
                  ],
                  "name": "Foo",
                  "qualifier": "TRK",
                }
              }
            />
          }
          right={
            <div
              className="display-flex-center"
            >
              <React.Fragment>
                <div
                  id="measures-view-selection-label"
                >
                  component_measures.view_as
                </div>
                <MeasureViewSelect
                  className="measure-view-select spacer-left big-spacer-right"
                  handleViewChange={[Function]}
                  metric={
                    {
                      "direction": 1,
                      "domain": "Reliability",
                      "id": "1",
                      "key": "bugs",
                      "name": "Bugs",
                      "type": "INT",
                    }
                  }
                  view="list"
                />
                <PageActions
                  componentQualifier="TRK"
                  showShortcuts={true}
                  total={0}
                />
              </React.Fragment>
            </div>
          }
        />
      </div>
    </div>
  </div>
  <div
    className="layout-page-main-inner measure-details-content"
  >
    <MeasureHeader
      component={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      measureValue="12"
      metric={
        {
          "direction": 1,
          "domain": "Reliability",
          "id": "1",
          "key": "bugs",
          "name": "Bugs",
          "type": "INT",
        }
      }
    />
    <FilesView
      components={[]}
      defaultShowBestMeasures={false}
      fetchMore={[Function]}
      handleOpen={[Function]}
      handleSelect={[Function]}
      loadingMore={true}
      metric={
        {
          "direction": 1,
          "domain": "Reliability",
          "id": "1",
          "key": "bugs",
          "name": "Bugs",
          "type": "INT",
        }
      }
      metrics={
        {
          "bugs": {
            "direction": 1,
            "domain": "Reliability",
            "id": "1",
            "key": "bugs",
            "name": "Bugs",
            "type": "INT",
          },
        }
      }
      paging={
        {
          "pageIndex": 12,
          "pageSize": 500,
          "total": 0,
        }
      }
      rootComponent={
        {
          "key": "foo",
          "measures": [
            {
              "bestValue": false,
              "metric": "bugs",
              "value": "12",
            },
          ],
          "name": "Foo",
          "qualifier": "TRK",
        }
      }
      view="list"
    />
  </div>
</div>
`;
